File: app/src/main/java/com/weatherwise/viewmodel/WeatherViewModel.java
package com.weatherwise.viewmodel;

import android.app.Application;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

import com.weatherwise.api.WeatherRepository;
import com.weatherwise.model.CurrentWeather;
import com.weatherwise.model.OneCallResponse;
import com.weatherwise.util.AppConstants;

public class WeatherViewModel extends AndroidViewModel {

    private static final String PREF_NAME       = "weatherwise_prefs";
    private static final String PREF_LAST_CITY  = "last_city";
    private static final String PREF_UNITS      = "units";

    private final WeatherRepository repository;
    private final SharedPreferences prefs;

    // ── LiveData yang diekspos ke UI ───────────────────────────
    public final LiveData<CurrentWeather>  currentWeather;
    public final LiveData<OneCallResponse> oneCallData;
    public final LiveData<Boolean>         isLoading;
    public final LiveData<String>          errorMessage;
    public final LiveData<String>          cityName;

    // MediatorLiveData — gabungkan currentWeather + oneCallData
    // agar UI tahu kapan kedua data sudah siap
    private final MediatorLiveData<Boolean> _isDataReady = new MediatorLiveData<>();
    public  final LiveData<Boolean>          isDataReady  = _isDataReady;

    public WeatherViewModel(@NonNull Application application) {
        super(application);

        repository = new WeatherRepository();
        prefs      = application.getSharedPreferences(PREF_NAME,
                android.content.Context.MODE_PRIVATE);
        repository.setUnits(prefs.getString(PREF_UNITS, "metric"));

        // Sambungkan LiveData dari Repository
        currentWeather = repository.currentWeather;
        oneCallData    = repository.oneCallData;
        isLoading      = repository.isLoading;
        errorMessage   = repository.errorMessage;
        cityName       = repository.currentCityName;

        // MediatorLiveData: isDataReady = true jika keduanya tidak null
        _isDataReady.addSource(currentWeather, cw ->
                _isDataReady.setValue(cw != null && oneCallData.getValue() != null)
        );
        _isDataReady.addSource(oneCallData, oc ->
                _isDataReady.setValue(currentWeather.getValue() != null && oc != null)
        );

        // Auto-load kota terakhir saat ViewModel dibuat
        loadLastCity();
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Cari cuaca berdasarkan nama kota
    // ──────────────────────────────────────────────────────────
    public void searchByCity(String cityName) {
        if (cityName == null || cityName.trim().isEmpty()) return;

        String trimmed = cityName.trim();
        saveLastCity(trimmed); // simpan untuk sesi berikutnya
        repository.fetchWeatherByCity(trimmed);
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Cari cuaca berdasarkan koordinat GPS
    // ──────────────────────────────────────────────────────────
    public void searchByCoord(double lat, double lon) {
        repository.fetchWeatherByCoord(lat, lon);
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Refresh data dengan koordinat terakhir
    // Dipanggil oleh pull-to-refresh di HomeScreen
    // ──────────────────────────────────────────────────────────
    public void refresh() {
        repository.refresh();
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Hapus pesan error
    // Dipanggil saat user dismiss ErrorBanner
    // ──────────────────────────────────────────────────────────
    public void clearError() {
        repository.clearError();
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Toggle satuan suhu Celsius ↔ Fahrenheit
    // ──────────────────────────────────────────────────────────
    public void toggleUnits() {
        String current = prefs.getString(PREF_UNITS, "metric");
        String next    = current.equals("metric") ? "imperial" : "metric";
        prefs.edit().putString(PREF_UNITS, next).apply();

        repository.setUnits(next);

        // Refresh data dengan satuan baru
        repository.refresh();
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Getter satuan saat ini
    // ──────────────────────────────────────────────────────────
    public boolean isCelsius() {
        return prefs.getString(PREF_UNITS, "metric").equals("metric");
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Ambil alert cuaca dari OneCallResponse (bisa null)
    // ──────────────────────────────────────────────────────────
    public java.util.List<OneCallResponse.AlertData> getAlerts() {
        OneCallResponse oneCall = oneCallData.getValue();
        if (oneCall == null || oneCall.alerts == null) {
            return new java.util.ArrayList<>();
        }
        return oneCall.alerts;
    }

    // ──────────────────────────────────────────────────────────
    // PUBLIC: Ambil data daily forecast (maks 7 hari)
    // Dipanggil oleh ForecastScreen
    // ──────────────────────────────────────────────────────────
    public java.util.List<OneCallResponse.DailyData> getDailyForecast() {
        OneCallResponse oneCall = oneCallData.getValue();
        if (oneCall == null || oneCall.daily == null) {
            return new java.util.ArrayList<>();
        }
        // Lewati hari ini (index 0), tampilkan 7 hari ke depan
        return oneCall.daily.size() > 1
                ? oneCall.daily.subList(1, Math.min(8, oneCall.daily.size()))
                : new java.util.ArrayList<>();
    }

    // ──────────────────────────────────────────────────────────
    // PRIVATE: Load kota terakhir dari SharedPreferences
    // ──────────────────────────────────────────────────────────
    private void loadLastCity() {
        String lastCity = prefs.getString(PREF_LAST_CITY, AppConstants.DEFAULT_CITY);
        repository.fetchWeatherByCity(lastCity);
    }

    // ──────────────────────────────────────────────────────────
    // PRIVATE: Simpan kota terakhir ke SharedPreferences
    // ──────────────────────────────────────────────────────────
    private void saveLastCity(String cityName) {
        prefs.edit().putString(PREF_LAST_CITY, cityName).apply();
    }
}

File: app/src/main/java/com/weatherwise/viewmodel/WeatherViewModel.kt
package com.weatherwise.viewmodel

import android.app.Application
import android.content.SharedPreferences
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.weatherwise.api.WeatherRepository
import com.weatherwise.model.GeocodingResponse
import com.weatherwise.model.WeatherAlert
import com.weatherwise.util.AppConstants
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

@OptIn(FlowPreview::class)
class WeatherViewModel(application: Application) : AndroidViewModel(application) {
    private val prefs: SharedPreferences = application.getSharedPreferences("weatherwise_prefs", android.content.Context.MODE_PRIVATE)
    private val repository = WeatherRepository(application)

    val currentWeather = repository.currentWeather
    val hourlyForecast = repository.hourlyForecast
    val dailyForecast = repository.dailyForecast
    val isLoading = repository.isLoading
    val errorMessage = repository.errorMessage
    val cityName = repository.currentCityName

    // Search & Autocomplete State
    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _searchSuggestions = MutableStateFlow<List<GeocodingResponse>>(emptyList())
    val searchSuggestions: StateFlow<List<GeocodingResponse>> = _searchSuggestions.asStateFlow()

    // Weather Alerts State
    private val _weatherAlerts = MutableStateFlow<List<WeatherAlert>>(emptyList())
    val weatherAlerts: StateFlow<List<WeatherAlert>> = _weatherAlerts.asStateFlow()

    private val _isDataReady = MutableStateFlow(false)
    val isDataReady: StateFlow<Boolean> = _isDataReady.asStateFlow()

    init {
        repository.setUnits(prefs.getString("units", "metric") ?: "metric")
        
        // Load cached data first for instant UI
        viewModelScope.launch {
            repository.loadCachedWeather()
        }
        
        // If no cache, load last city
        viewModelScope.launch {
            val lastCity = prefs.getString("last_city", AppConstants.DEFAULT_CITY) ?: AppConstants.DEFAULT_CITY
            if (repository.currentWeather.value == null) {
                repository.fetchWeatherByCity(lastCity)
            }
        }

        // Setup search debounce
        _searchQuery
            .debounce(500)
            .filter { it.trim().length >= 2 }
            .distinctUntilChanged()
            .onEach { query ->
                fetchGeocodingSuggestions(query.trim())
            }
            .launchIn(viewModelScope)
    }

    private suspend fun fetchGeocodingSuggestions(query: String) {
        try {
            val suggestions = repository.fetchGeocoding(query)
            _searchSuggestions.value = suggestions
        } catch (e: Exception) {
            _searchSuggestions.value = emptyList()
        }
    }

    fun updateSearchQuery(query: String) {
        _searchQuery.value = query
        if (query.isBlank()) {
            _searchSuggestions.value = emptyList()
        }
    }

    fun selectSuggestion(suggestion: GeocodingResponse) {
        _searchQuery.value = suggestion.name
        _searchSuggestions.value = emptyList()
        searchByCoord(suggestion.lat, suggestion.lon)
    }

    fun searchByCity(cityName: String) {
        if (cityName.isBlank()) return
        _searchQuery.value = ""
        _searchSuggestions.value = emptyList()
        prefs.edit().putString("last_city", cityName.trim()).apply()
        viewModelScope.launch {
            repository.fetchWeatherByCity(cityName.trim())
        }
    }

    fun searchByCoord(lat: Double, lon: Double) {
        _searchQuery.value = ""
        _searchSuggestions.value = emptyList()
        viewModelScope.launch {
            repository.fetchWeatherByCoord(lat, lon)
        }
    }

    fun refresh() {
        viewModelScope.launch {
            repository.refresh()
        }
    }

    fun clearError() {
        repository.clearError()
    }

    fun toggleUnits() {
        val current = prefs.getString("units", "metric") ?: "metric"
        val next = if (current == "metric") "imperial" else "metric"
        prefs.edit().putString("units", next).apply()
        repository.setUnits(next)
        viewModelScope.launch {
            repository.refresh()
        }
    }

    fun isCelsius(): Boolean = (prefs.getString("units", "metric") ?: "metric") == "metric"
}